clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -307,13 +307,26 @@ public long timeToNextPoll(long now) {
         client.ensureFreshMetadata();
 
         isLeader = true;
-        assignmentSnapshot = metadataSnapshot;
 
         log.debug("Performing assignment for group {} using strategy {} with subscriptions {}",
                 groupId, assignor.name(), subscriptions);
 
         Map<String, Assignment> assignment = assignor.assign(metadata.fetch(), subscriptions);
 
+        // check if the assignor has realized some new topics, if yes update the snapshot
```

Comment:
@hachikuji This is just to show my current thoughts about how to work around the unnecessary rebalance without changing `Assignor` APIs, this is really not that elegant but just as a demo, and I'm happy to revert this back for future general solution; my take is that if we would allow `assign()` functions to update the `metadatasnapshot` then we probably would need to change the API.